### DISCUSSION TITLE
Adaptation to Arduino IDE 2.2

### DISCUSSION THREAD

**Discussion by TPlattner** [pschatzmann/TinyFTPClient#2] [Q&A] [answered]
Since the release of the Arduino IDE 2.2, it is recommended for the ESP8266 to use the new file "ESP8266WiFi.h" with new and improved functions instead of "WiFi.h" to use the WLAN.
However, "TinyFTPClient.h" uses "WiFi.h" as an include, which leads to numerous warnings for me in the Arduino IDE 2.2.1 in contrast to "ESP8266WiFi.h".
Would it be possible to convert "TinyFTPClient.h" specifically for the ESP8266 so that the "ESP8266WiFi.h" is used?

**Comment by pschatzmann** [comment, marked as answer]
I have updated my library yesterday and removed all platform specific logic. 

I suggest to try the latest version and do not forget to read the updated README of this project since there are some changes in the initialization.

**Reply by TPlattner** [reply]
Great, it basically works now.
I have made a change in the file "Tiny_FTPClient.h" in line 65 and inserted an "unsigned" before "int i = 0". This causes the program to compile without a warning.
However, there is a problem when downloading a file that is larger than the configured buffer. If the buffer area is full and the data is written to a local file, characters are added and the connection is aborted. In addition, the file becomes longer, 175 characters are read and 200 characters are saved.

**Comment by pschatzmann** [comment]
Can you share your code: you must be doing something wrong. 
I was downloading mp3 files w/o loosing any information

    byte buffer[100];
    while (file.available()>0){
        int len = file.read(buffer, 100);
        Serial.write(buffer, len);
    }

**Reply by TPlattner** [reply]
`
void` Daten_holen(uint16_t year)
{
  const char* filepraefix = "Abfall_";
  const char* filesuffix = ".csv";
  char* filename = new char[20];
  unsigned char filebuffer[256];

  sprintf(filename, "%s%d%s", filepraefix, (year + 2000), filesuffix);
  client.begin(IPAddress(192,168,20,1), ftp_user, ftp_pw);

  LittleFS.format();
  LittleFS.begin();
  if (LittleFS.exists(filename) == true) LittleFS.remove(filename);
  File esp_file = LittleFS.open(filename, "a");

  // optional logging
    FTPLogger::setOutput(Serial);
    FTPLogger::setLogLevel(LOG_DEBUG);

  // copy data from file
  FTPFile file = client.open(filename);

  int len;
  while (file.available()>0){
      len = file.read(filebuffer, sizeof(filebuffer));
      esp_file.write(filebuffer, len);
      Serial.print((char*)filebuffer);
      //delay(1000);
  }

  Serial.print("Dateigröße:  ");
  Serial.print(esp_file.size());
  Serial.println(" Bytes");
  esp_file.close();
  client.end();
`

**Reply by TPlattner** [reply]
This works, the buffer is larger than the file. If the file is larger than the buffer, the output looks like this:

FTP DEBUG - FTPBasicAPI::cmd RETR Abfall_2023.csv
FTP DEBUG - FTPBasicAPI::checkResult 425 Can't build data connection: Connection refused.

FTP DEBUG - FTPBasicAPI::checkResult - checking with 150
FTP DEBUG - FTPBasicAPI::checkResult - checking with 125
FTP ERROR - FTPBasicAPI::checkResult RETR Abfall_2023.csv
FTP ERROR - FTPBasicAPI::checkResult 425 Can't build data connection: Connection refused.

FTP DEBUG - FTPBasicAPI checkClosed -> client is closed
FTP DEBUG - FTPFile available: 136
FTP INFO - FTPFile read-n
FTP DEBUG - FTPBasicAPI read
FTP DEBUG - FTPBasicAPI::cmd RETR Abfall_2023.csv
FTP DEBUG - FTPBasicAPI::checkResult 425 Can't build data connection: Connection refused.

FTP DEBUG - FTPBasicAPI::checkResult - checking with 150
FTP DEBUG - FTPBasicAPI::checkResult - checking with 125
FTP ERROR - FTPBasicAPI::checkResult RETR Abfall_2023.csv
FTP ERROR - FTPBasicAPI::checkResult 425 Can't build data connection: Connection refused.

FTP DEBUG - FTPBasicAPI checkClosed -> client is closed
306;1

310;2

312;1

313;8

319;1

324;6

327;1

328;8

@� @���?�?L� @FTP DEBUG - FTPBasicAPI read
FTP DEBUG - FTPBasicAPI::cmd RETR Abfall_2023.csv
FTP DEBUG - FTPBasicAPI::checkResult 425 Can't build data connection: Connection refused.

FTP DEBUG - FTPBasicAPI::checkResult - checking with 150
FTP DEBUG - FTPBasicAPI::checkResult - checking with 125
FTP ERROR -

**Comment by pschatzmann** [comment]
I don't think this has anything to do with any buffers but seem rather an [infrastructure problem](https://www.google.com/search?q=425+Can%27t+build+data+connection%3A+Connection+refused.&oq=425+Can%27t+build+data+connection%3A+Connection+refused.&gs_lcrp=EgZjaHJvbWUyBggAEEUYOdIBBzU0NGowajeoAgCwAgA&sourceid=chrome&ie=UTF-8).

What OS are you using ? If you are using windows, I can't really help you...

**Reply by TPlattner** [reply]
The ftp-server is an AVM-Fritzbox 6440 with operating system version 7.57.
No Windows computers are involved in the transfer.
Can you give me a tip on how to change the transfer mode from ASCII to BINARY? It would be nice if this parameter could be specified when opening the connection to the ftp server.

**Reply by pschatzmann** [reply]
Oh, my ftp server is binary by default. Is yours really starting in ascii ? 
I have added the ascii() and binary() commands and select binary on startup now...

**Reply by TPlattner** [reply]
I have not found a way to set a default value for the transfer mode in the Fritzbox.
By default, the transfer is still carried out as ASCII.
Which command do I have to insert where in order to force a transfer as binary?

from protocol:
connecting to 192.168.20.1
FTP INFO - FTPClient open
FTP DEBUG - FTPFile
FTP DEBUG - FTPBasicAPI read
FTP DEBUG - FTPBasicAPI::cmd RETR Abfall_2023.csv
FTP DEBUG - FTPBasicAPI::checkResult 150 Opening ASCII mode data connection for 'Abfall_2023.csv' (175 bytes).

FTP DEBUG - FTPBasicAPI::checkResult - checking with 150
FTP DEBUG - FTPBasicAPI::checkResult  -> success with 150
FTP DEBUG - FTPBasicAPI checkClosed -> client is closed
FTP DEBUG - FTPFile available: 200
FTP INFO - FTPFile read-n

**Reply by pschatzmann** [reply]
With the new version I am automatically calling the binary() method on startup.

**Reply by TPlattner** [reply]
However, the ftp server of the Fritzbox does not seem to understand this correctly (line 6 of the protocol). There seems to be something wrong.  
With FileZilla I can select ASCII or BINARY and can also see the differences in the delivered file. In principle, the ftp server can do both.

**Comment by pschatzmann** [comment]
What response do you get when you login to the ftp server with the ftp command and type the command bin ?
What is the result when you enter Serial.println(client.binary() ? "bin OK" : "bin Error")); in your sketch ?

**Reply by TPlattner** [reply]
The line
**Serial.println(client.binary() ? "bin OK" : "bin Error"));**
generates an error: 'class FTPClient' has no member named 'binary'

Here is the log of a login to the ftp server:
Status:	Verbinde mit 192.168.20.1:21...
Status:	Verbindung hergestellt, warte auf Willkommensnachricht...
Antwort:	220 FTP server ready.
Befehl:	AUTH TLS
Antwort:	234 Authentication method accepted
Status:	Initialisiere TLS...
Status:	TLS-Verbindung hergestellt.
Befehl:	USER ftp_abfall
Antwort:	331 Password required for ftp_abfall.
Befehl:	PASS ****************
Antwort:	230 User ftp_abfall logged in.
Befehl:	OPTS UTF8 ON
Antwort:	200 UTF8 ON
Befehl:	PBSZ 0
Antwort:	200 ok
Befehl:	PROT P
Antwort:	200 Data channel will be secured
Status:	Angemeldet
Status:	Empfange Verzeichnisinhalt...
Befehl:	PWD
Antwort:	257 "/" is current directory.
Befehl:	TYPE I
Antwort:	200 Type set to I.
Befehl:	PASV
Antwort:	227 Entering Passive Mode (192,168,20,1,211,33)
Befehl:	LIST
Antwort:	150 Opening BINARY mode data connection for '/bin/ls -lgA'.
Antwort:	226 Transfer complete.
Status:	Anzeigen des Verzeichnisinhalts für "/" abgeschlossen
Befehl:	help
Antwort:	214- The following commands are recognized (* =>'s unimplemented).
Antwort:	   USER    PASV    MLFL*   ALLO    XCWD    NOOP    XPWD    EPRT 
Antwort:	   PASS    TYPE    MAIL*   REST    LIST    MKD     CDUP    FEAT 
Antwort:	   ACCT*   STRU    MSND*   RNFR    NLST    XMKD    XCUP    OPTS 
Antwort:	   SMNT*   MODE    MSOM*   RNTO    SITE    RMD     STOU    AUTH 
Antwort:	   REIN*   RETR    MSAM*   ABOR    SYST    XRMD    SIZE    PBSZ 
Antwort:	   QUIT    STOR    MRSQ*   DELE    STAT    RRMD    MDTM    PROT 
Antwort:	   PORT    APPE    MRCP*   CWD     HELP    PWD     EPSV 
Antwort:	214 Direct comments to support@avm.de.
Befehl:	type
Antwort:	500 'TYPE': command not understood.
The commands "BIN" or "BINARY" are not implemented (command not understood), but the implemented command "TYPE" is also answered with "command not understood".


Here is the log from loading a file for display in Filezilla. It shows the BINARY mode.
However, the TinyFTP client reports ASCII mode when the ESP triggers the download.
Status:	Verbinde mit 192.168.20.1:21...
Status:	Verbindung hergestellt, warte auf Willkommensnachricht...
Antwort:	220 FTP server ready.
Befehl:	AUTH TLS
Antwort:	234 Authentication method accepted
Status:	Initialisiere TLS...
Status:	TLS-Verbindung hergestellt.
Befehl:	USER ftp_abfall
Antwort:	331 Password required for ftp_abfall.
Befehl:	PASS ****************
Antwort:	230 User ftp_abfall logged in.
Befehl:	OPTS UTF8 ON
Antwort:	200 UTF8 ON
Befehl:	PBSZ 0
Antwort:	200 ok
Befehl:	PROT P
Antwort:	200 Data channel will be secured
Status:	Angemeldet
Status:	Starte Download von /Abfall_2023.csv
Befehl:	CWD /
Antwort:	250 CWD command successful.
Befehl:	PWD
Antwort:	257 "/" is current directory.
Status:	Empfange Verzeichnisinhalt für "/"...
Befehl:	TYPE I
Antwort:	200 Type set to I.
Befehl:	PASV
Antwort:	227 Entering Passive Mode (192,168,20,1,220,178)
Befehl:	LIST
Antwort:	150 Opening BINARY mode data connection for '/bin/ls -lgA'.
Antwort:	226 Transfer complete.
Befehl:	PASV
Antwort:	227 Entering Passive Mode (192,168,20,1,212,156)
Befehl:	RETR Abfall_2023.csv
Antwort:	150 Opening BINARY mode data connection for 'Abfall_2023.csv' (175 bytes).
Antwort:	226 Transfer complete.
Status:	Dateiübertragung erfolgreich, 175 Bytes in 1 Sekunde übertragen

I don't know what to do here either.

**Comment by pschatzmann** [comment]
I did not know that there are such big differences!

Last trial: I added the type command: so client.type("I") should give binary mode in your server. For me this is not working on my server.
I suggest to check in your sketch if this command returns true 

ps and don't forget to install the latest library version.

**Reply by TPlattner** [reply]
The last commit of ArduinoFTPClient.h and .cpp was 8:27 AM, did you upload a new version?
Compiling message:
C:\Users\Thomas\Documents\Arduino\GxEPD2_ESP8266_Abfall\GxEPD2_ESP8266_Abfall.ino: In function 'void Daten_holen(uint16_t)':
C:\Users\Thomas\Documents\Arduino\GxEPD2_ESP8266_Abfall\GxEPD2_ESP8266_Abfall.ino:275:25: error: 'class FTPClient' has no member named 'binary'
  275 |   Serial.println(client.binary() ? "bin OK" : "bin Error"));
      |                         ^~~~~~
C:\Users\Thomas\Documents\Arduino\GxEPD2_ESP8266_Abfall\GxEPD2_ESP8266_Abfall.ino:276:10: error: 'class FTPClient' has no member named 'type'
  276 |   client.type("I");
      |          ^~~~

**Reply by pschatzmann** [reply]
You will need to download the main branch! 
You can easily double check by looking at the ArduinoFTPClient.h in the src directory on Github and compare them with what you have in your libraries folder.

The latest version will have the binary() and type() methods...

**Reply by TPlattner** [reply]
Please excuse my stupid questions, but for me these are the first steps in C++ and the class system.
I think I am using the right files. To be safe, I copied the contents of the files directly into the files in the library directory and did not use the installation via the downloaded ZIP file.
In ArduinoFTPClient.h I find the declaration:
```
class FTPBasicAPI {
  friend class FTPFile;

public:
  FTPBasicAPI();
  ~FTPBasicAPI();
  ...
  virtual bool binary();
  virtual bool ascii();
  virtual bool type(const char *type);

And the corresponding definitions in the ArduinoFTPClient.cpp file:
bool FTPBasicAPI::binary() { ...

bool FTPBasicAPI::ascii() { ... }

bool FTPBasicAPI::type(const char* txt) { ...
```

In the sketch  after:
`FTPClient client(cmd, data);`
the functions are called with 
```
Serial.println(client.binary() ? "bin OK" : "bin Error"));
client.type("I");
```
Is this really correct? I would assume that the errors during compilation occur because "type" and "binary" are members of the FTPClient class and do not belong to the "FTPBasicAPI" class.
But as I said, I'm a complete novice with C++.

**Comment by pschatzmann** [comment]
Oh, yes...you are absolutely right!

**Comment by pschatzmann** [comment]
I added the [methods](https://pschatzmann.github.io/TinyFTPClient/docs/html/class_f_t_p_client.html) to the client as well and now your compile errors should be resolved.

Do not use the binary() method on your server, since this is not supported.
I would do something as follows:

    if (!client.type("I")){
       Serial.println("Could not set binary mode");
       while(true); // stop processing
    }

**Reply by TPlattner** [reply]
I can now transfer files in binary mode and both files have the same length.
Thank you very much for your support.

**Comment by pschatzmann** [comment]
Cool!
